doc/fluid/design/imperative/imperative.md

```diff
@@ -0,0 +1,111 @@
+# Imperative Programming
+
+This document will discuss changes needed in Paddle platform to support
+imperative programming.  
+
+Paddle provides a Python API that users typically write their training or
+inference programs in.  Paddle will convert this program into a ProgramDesc, 
+which is later executed in the C++ runtime.
+
+With this flexibility, a Paddle program could contain a mix of pure Python API
+and Paddle Fluid API.  To further complicate the code, the neural network logic
+could be combined with the run logic (ie: Logic to run on a distributed 
+system).
+
+The mix of Python and Paddle API also makes it impossible for a transpiler to
+convert the ProgramDesc to C++.
+
+## Changes to support Imperative
+
+Please refer to [Paddle API v4 Proposal](https://github.com/PaddlePaddle/Paddle/issues/9912)
```

Comment:
This is not true imperative because we do not have eager execution.  This change is about moving the all Computation Representation to the IR (ProgramDesc).  Currently users mix in Python code with IR (ie: load training data in python, run executor manually for startup and main program), which makes it difficult to transpile.